### PR TITLE
Health check support in TChannelAsThrift

### PR DESCRIPTION
**Pull request by ShanniLi**
r @Raynos @kriskowal @anson627

### CODE REVIEW

File: node/as/thrift.js [hunk 1]
```diff
@@ -53,8 +55,29 @@ function TChannelAsThrift(opts) {
     self.logParseFailures = typeof logParseFailures === 'boolean' ?
         logParseFailures : true;
 
-    // only used in request()
     self.channel = opts.channel;
+
+    self.healthCheckCallback = opts.healthCheckCallback;
+    assert(!self.healthCheckCallback || typeof self.healthCheckCallback === 'function',
```
Comment (Raynos) [inline node/as/thrift.js]: healthCheckCallback must be optional for back compat.

File: node/as/thrift.js [hunk 2]
```diff
@@ -53,8 +55,29 @@ function TChannelAsThrift(opts) {
     self.logParseFailures = typeof logParseFailures === 'boolean' ?
         logParseFailures : true;
 
-    // only used in request()
     self.channel = opts.channel;
+
+    self.healthCheckCallback = opts.healthCheckCallback;
```
Comment (Raynos) [inline node/as/thrift.js]: Let's call this `self.isHealthy = opts.isHealthy`. This is not a callback so cant be called Callback.

File: node/as/thrift.js (hunk 1 above)
Comment (Raynos) [inline node/as/thrift.js]: Oh I see; got it, cool :)

File: node/as/thrift.js [hunk 3]
```diff
@@ -65,6 +88,17 @@ function TChannelThriftRequest(options) {
     self.tchannelThrift = options.tchannelThrift;
 }
 
+function health(self, req, head, body, callback) {
+    var status = self.healthCheckCallback();
```
Comment (Raynos) [inline node/as/thrift.js]: Let's add `assert('ok' in status, 'Status must have ok field')` and `assert('message' in status, 'Status must have message field')`

**Comment by Raynos** [comment]
Please also update the docs folder.

**Comment by ShanniLi** [comment]
@Raynos confused. which doc file?

File: node/as/thrift.js (hunk 3 above)
Comment (ShanniLi) [inline node/as/thrift.js]: Will add ok. However, message is optional in thrift.

File: node/as/thrift.js [hunk 4]
```diff
@@ -65,6 +88,17 @@ function TChannelThriftRequest(options) {
...
+    return callback(null, {
```
Comment (anson627) [inline node/as/thrift.js]: return sth return by a callback passed in by caller? what's the use case here?

File: node/as/thrift.js (hunk 3 above)
Comment (Raynos) [inline node/as/thrift.js]: Good point; `assert(!status.ok || 'message' in status)`

**Comment by Raynos** [comment]
@ShanniLi docs/thrift.md

File: node/as/thrift.js (hunk 4 above)
Comment (ShanniLi) [inline node/as/thrift.js]: used to provide a health check wrapper for your service, so your service don't have to worry about meta.thrift, etc.

**Comment by anson627** [comment]
rebase master to pass Jenkins? otherwise lgtm

File: node/as/thrift.js (hunk 4 above)
Comment (anson627) [inline node/as/thrift.js]: I see. The doc make it clear.

File: node/as/thrift.js [hunk 5]
```diff
@@ -65,6 +88,21 @@ function TChannelThriftRequest(options) {
     self.tchannelThrift = options.tchannelThrift;
 }
 
+function health(self, req, head, body, callback) {
+    var status = self.healthCheckCallback();
+    assert(status && typeof status.ok === 'boolean', 'status must have ok field');
+    assert(status && (!status.message || typeof status.message === 'string'),
```
Comment (Raynos) [inline node/as/thrift.js]: if status.ok === false then there must be a message.

File: node/docs/as-thrift.md [hunk 6]
```diff
@@ -91,7 +91,13 @@ It can be passed options.
     `logger.warn()`. If you do not want these log statements you
     can set the option to `false`
  - `opts.channel` The channel used to make requests on. This
-    option is required if you want to use the `.request()` method
+    option is required if you want to use the `.request()` method or
+    `opts.healthCheckCallback` is passed in.
+ - `opts.healthCheckCallback` The callback function used to return the
```
Comment (Raynos) [inline node/docs/as-thrift.md]: Let's pick a better name. This is not a callback. Consider `isHealthy`

File: node/test/health.js [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+    var client = cluster.channels[1].subChannels.server;
+
+    tchannelAsThrift.send(client.request({
+        serviceName: 'server',
+        hasNoParent: true
+    }), 'Chamber::echo', null, {
+        value: 10
+    }, function onResponse(err, res) {
+        assert.ifError(err);
```
Comment (kriskowal) [inline node/test/health.js]: please use the `if (err) return assert.end(err)` pattern. Failing tests should not crash the test runner.

File: node/test/health.js [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        source: source,
+        logParseFailures: false
+    });
+
+    healthThrift.send(client.request({
+        serviceName: 'server',
+        hasNoParent: true
+    }), 'Meta::health', null, {}, function onResponse(err, res) {
+        if (err) {
+            assert.end(false);
```
Comment (kriskowal) [inline node/test/health.js]: `assert.end(err)`.

File: node/test/health.js [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    });
+
+    var options = {
+        isOptions: true
+    };
+
+    var health = opts.good ? healthGood : healthBad;
+
+    var tchannelAsThrift = new TChannelAsThrift({
+        source: opts.thriftText || globalThriftText,
```
Comment (kriskowal) [inline node/test/health.js]: Please name this `thriftSource` or simply `source`. Don’t mix `text` and `source` for the same semantics.

File: node/as/thrift.js [hunk 10]
```diff
@@ -53,8 +55,29 @@ function TChannelAsThrift(opts) {
...
         logParseFailures : true;
 
-    // only used in request()
     self.channel = opts.channel;
+
+    self.healthCheckCallback = opts.healthCheckCallback;
+    assert(!self.healthCheckCallback || typeof self.healthCheckCallback === 'function',
+        'healthCheckCallback must be a function');
+    assert(!self.healthCheckCallback || self.channel,
+        'channel must be provided with healthCheckCallback');
```
Comment (kriskowal) [inline node/as/thrift.js]: Let’s be sure to note that this is a backward incompatible change in the migration docs, and make a major version bump, or not make a breaking change.

Comment (ShanniLi) [inline node/as/thrift.js]: it is actually backward compatible

**Comment by Raynos** [comment]
lgtm.

**Pull request merged**